packages/labs/router/src/routes.ts

```diff
@@ -163,14 +167,24 @@ export class Routes implements ReactiveController {
       // Simulate a tail group with the whole pathname
       this._currentParams = {0: tailGroup};
     } else {
-      const route = (this._currentRoute = this._getRoute(pathname));
+      const route = this._getRoute(pathname);
       if (route === undefined) {
         throw new Error(`No route found for ${pathname}`);
       }
       const pattern = getPattern(route);
-      const r = pattern.exec({pathname});
-      this._currentParams = r?.pathname.groups ?? {};
-      tailGroup = getTailGroup(this._currentParams);
+      const result = pattern.exec({pathname});
+      const params = result?.pathname.groups ?? {};
+      tailGroup = getTailGroup(params);
+      if (typeof route.enter === 'function') {
+        const success = await route.enter(params);
```

Comment:
what happens if route.enter rejects?

Reply:
That should fine here since none of the state for matching this route is recorded unless enter() returns non-false. Obviously we need a lot more tests...

Reply:
I think `enter()` (and maybe an `exit()`) is good because it lets `render()` stay synchronous, can reject a route transition, and signals that the route is in a pending state, which we might want to expose to `render()` so that it can rendering a loading message.

We also had a request to add middleware to the router. I need to go back and check the examples, but it seemed like middleware might be able to just hook enter/exit and render is still just part of the route configuration.

Reply:
> We also had a request to add middleware to the router.

https://lit-and-friends.slack.com/archives/C03PF4L4L/p1638556109035200?thread_ts=1638548761.018500&cid=C03PF4L4L

Reply:
omg, thanks for adding that link. I was just about to ask you where that was :)

Reply:
Some more context:

Here's a more up to date example. Arguably some of these things could/should even be baked into the router, like resetting focus and setting the page title, and would be configurable in the router configuration

```js
const routes = [
  {
    path: '/shop',
    component: 'shop-page',
    action: middleware(
      title('shop'), // sets page title
      offline, // redirects user to the offline screen
      auth, // checks auth, if not authed redirect to offline screen
      lazy(() => import('../pages/shop-page/index.js')), // lazy load
      scrollIntoView, // scrolls a position
      restoreFocus // resets focus
    )
  },
];
```

Here's the code for the middleware function, its not very big:

```js
async function invokeMiddlewares(middlewares) {
  if (!middlewares.length) return;

  const mw = middlewares[0];

  return mw(async () => {
    await invokeMiddlewares(middlewares.slice(1));
  });
}

export function middleware(...middlewares) {
  return () => {
    invokeMiddlewares(middlewares);
  };
}
```

Reply:
Regarding the title update...
I think we will want to update more things like the description and the meta tags.